src/SFML/Graphics/Shader.cpp

```diff
@@ -40,6 +40,18 @@
 
 #ifndef SFML_OPENGL_ES
 
+#if defined(SFML_SYSTEM_MACOS) || defined(SFML_SYSTEM_IOS)
+
+    #define castToGlHandle(x) reinterpret_cast<GLEXT_GLhandle>(static_cast<ptrdiff_t>(x))
+    #define castFromGlHandle(x) static_cast<unsigned int>(reinterpret_cast<ptrdiff_t>(x))
+
+#else
+
+    #define castToGlHandle(x) (x)
+    #define castFromGlHandle(x) (x)
+
+#endif
```

Comment:
> It does. Since T is unsigned int, the following happens:

Tank provided an example for `castFromGlHandle()` but named it wrong and had the wrong return type (should be `unsigned int`).

You guys forget that this is the `castToGlHandle()` function. :wink: It has to cast from a fixed type (`unsigned int`) in both cases to whatever `GLEXT_GLhandle` is. `castFromGlHandle()` casts from a `GLEXT_GLhandle` to `unsigned int` in both cases. So yeah... your template solution only works for one direction but not the other.

> Actually, do you even need the case differentiation at all? Can't you just use reinterpret_cast() in all cases, even if input and output are identical? Never tried that to be honest, but don't see why it shouldn't work.

reinterpret_cast() doesn't allow casting from a pointer to an integral value when the conversion would truncate the data (i.e. 64-bit pointer -> 32-bit `unsigned int`). I tried this already. Apple really tried their hardest to be as annoying as possible. :wink: But yeah... I guess the inner static_cast in the "To" case can be removed. But moving the double cast into every line where it is needed would make the code really horrible to read IMHO.

I'm still interested to see suggestions for how to cast from an `unsigned int` to `GLEXT_GLhandle` that works in both Apple and non-Apple cases :grin:.